rhythm.py

```diff
@@ -1,24 +1,81 @@
+import logging
+from abc import ABCMeta, abstractmethod
+from time import sleep
+
 from regression import calculate_regression
-import math
 
-HANDSTROKE_GAP = 1
 
-class Rhythm:
-    logger_name = "RHYTHM"
+class Rhythm(metaclass=ABCMeta):
+    @abstractmethod
+    def wait_for_bell_time(self, current_time, bell, row_number, place, user_controlled, stroke):
+        pass
+
+    @abstractmethod
+    def expect_bell(self, expected_bell, row_number, index, expected_stroke):
+        pass
+
+    @abstractmethod
+    def on_bell_ring(self, bell, stroke, real_time):
+        pass
+
+    @abstractmethod
+    def initialise_line(self, stage, user_controls_treble, start_time):
+        pass
+
+
+class WaitForUserRhythm(Rhythm):
+    logger_name = "RHYTHM:WaitForUser"
+
+    def __init__(self, rhythm: Rhythm):
+        self._innerRhythm = rhythm
+        self._expected_bells = set()
+        self.delay = 0
+        self.logger = logging.getLogger(self.logger_name)
+
+    def wait_for_bell_time(self, current_time, bell, row_number, place, user_controlled, stroke):
+        self._innerRhythm.wait_for_bell_time(current_time - self.delay, bell, row_number, place, user_controlled, stroke)
+        if user_controlled:
+            delay_for_user = 0
+            while (bell, stroke) in self._expected_bells:
+                sleep(0.01)
+                delay_for_user += 0.01
+                self.logger.debug(f"Waiting for {bell}")
+            if delay_for_user:
+                self.logger.info(f"Delayed for {delay_for_user}")
+                self.delay += delay_for_user
+
+    def expect_bell(self, expected_bell, row_number, index, expected_stroke):
+        self._innerRhythm.expect_bell(expected_bell, row_number, index, expected_stroke)
+        self._expected_bells.add((expected_bell, expected_stroke))
+
+    def on_bell_ring(self, bell, stroke, real_time):
+        self._innerRhythm.on_bell_ring(bell, stroke, real_time - self.delay)
+        try:
+            self._expected_bells.remove((bell, stroke))
+        except KeyError:
+            pass
+
+    def initialise_line(self, stage, user_controls_treble, start_time):
+        self._innerRhythm.initialise_line(stage, user_controls_treble, start_time - self.delay)
+
+
+class RegressionRhythm(Rhythm):
```

Comment:
There may be much better names for this class to describe the algorithm you're implementing